common/src/main/java/io/crate/types/ObjectType.java

```diff
@@ -86,6 +88,24 @@ public DataType innerType(String key) {
         return innerTypes.getOrDefault(key, UndefinedType.INSTANCE);
     }
 
+    @Nullable
+    public DataType resolveInnerType(List<String> path) {
+        if (path.size() > 0) {
```

Comment:
Minor:

Could do an early `if (path.isEmpty() { return null; }` to reduce the indentation level of the rest of the function.

Another option would be to turn this into a non-recursive function similar to how `Maps.getByPath` is implemented. Would avoid the allocations caused by using `subList`

Reply:
+1 for an non-recursive function, will try to eliminate that.